File: Arrays and Hashing/118: Pascal's Triangle.js
// https://leetcode.com/problems/pascals-triangle/description/
// Difficulty: Easy

// Problem
/*
Given an integer numRows, return the first numRows of Pascal's triangle.
*/

// Solution
// O(n^2) time and O(1) auxillary space. Iterate over the previous row, and construct the new row. n^2 time since we have to process n rows of length n. Even though the rows are length 1, 2, ... n-1, n, it still works out to n^2.

var generate = function (numRows) {
  const result = [[1]];
  for (let i = 2; i <= numRows; i++) {
    const newRow = [];
    const prevRow = result[i - 2]; // [1]
    for (let j = 0; j < prevRow.length + 1; j++) {
      let newNum;
      if (j === 0 || j === prevRow.length) {
        newNum = 1;
      } else {
        newNum = prevRow[j] + prevRow[j - 1];
      }
      newRow.push(newNum);
    }
    result.push(newRow);
  }
  return result;
};

File: problems/Leetcode/118: Pascal's Triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        row = [1]
        res = [row]
        for _ in range(numRows - 1):
            nrow = [1]
            for i in range(len(res[-1]) - 1):
                v1 = res[-1][i]
                v2 = res[-1][i+1]
                nv = v1 + v2
                nrow.append(nv)
            nrow.append(1)
            res.append(nrow)
        return res

